src/XMLimport.cpp

```diff
@@ -375,133 +414,206 @@ void XMLimport::readAreas()
             {
-                pT->west = e;
+                pT->northeast = e;
+                if( ! entranceMultiHash.contains( e, pT->id ) )
+                {
+                    entranceMultiHash.insert( e, pT->id );
+                }
             }
-            if( dir == "east" )
+            else if( dir == QStringLiteral( "southwest" ) )
             {
-                pT->east = e;
+                pT->southwest = e;
+                if( ! entranceMultiHash.contains( e, pT->id ) )
+                {
+                    entranceMultiHash.insert( e, pT->id );
+                }
             }
-            if( dir == "up" )
+            else if( dir == QStringLiteral( "southeast" ) )
             {
-                pT->up = e;
+                pT->southeast = e;
+                if( ! entranceMultiHash.contains( e, pT->id ) )
+                {
+                    entranceMultiHash.insert( e, pT->id );
+                }
             }
-            if( dir == "down" )
+            else if( dir == QStringLiteral( "northwest" ) )
             {
-                pT->down = e;
+                pT->northwest = e;
+                if( ! entranceMultiHash.contains( e, pT->id ) )
+                {
+                    entranceMultiHash.insert( e, pT->id );
+                }
             }
-            if( dir == "in" )
+            else if( dir == QStringLiteral( "in" ) )
             {
                 pT->in = e;
+                if( ! entranceMultiHash.contains( e, pT->id ) )
+                {
+                    entranceMultiHash.insert( e, pT->id );
+                }
             }
-            if( dir == "out" )
+            else if( dir == QStringLiteral( "out" ) )
             {
                 pT->out = e;
+                if( ! entranceMultiHash.contains( e, pT->id ) )
+                {
+                    entranceMultiHash.insert( e, pT->id );
+                }
+            }
+            else
+            {
+                // TODO: Handle Special Exits
             }
             continue;
         }
+        else if( name() == QStringLiteral( "coord" ) )
+        {
+            if( attributes().value("x").toString().isEmpty() )
+            {
+                continue;
+            }
+
+            pT->x = attributes().value( QStringLiteral( "x" ) ).toString().toInt();
+            pT->y = attributes().value( QStringLiteral( "y" ) ).toString().toInt();
+            pT->z = attributes().value( QStringLiteral( "z" ) ).toString().toInt();
+            continue;
+        }
+        else if( Q_UNLIKELY( name().isEmpty() ) )
+        {
+            continue;
+        }
+
         if( isEndElement() )
         {
             break;
         }
     }
-    if( pT->id != 0 )
+
+    if( pT->id > 0 )
     {
+        mpHost->mpMap->reportStringToProgressDialog( tr( "Parsing Room data [id: %1]..." ).arg( pT->id ) );
+        areamRoomMultiHash.insert( pT->area, pT->id );
         mpHost->mpMap->mpRoomDB->addRoom( pT->id, pT );
```

Comment:
:snail: :hammer:  This makes all the difference - since we are loading/import a new map there is an optimisation that I missed enabling here - putting a true as a third argument:
`mpHost->mpMap->mpRoomDB->addRoom( pT->id, pT, true );`
 disables an unneeded step in adding a room { `TRoomDB::updateEntranceMap(TRoom *, bool)` can skip the computationally expensive `TRoomDB::deleteValuesFromEntranceMap( int )` step} this decimates (divides by 10?) the time to do each room and also pointed out quite clearly that the `TRoom::entranceMap` data was already being built and so there was no need to do it within `XMLimport::readRoom(...)` - I feel a bit of an idiot for missing that...